src/components/datatypeSelector/DatatypeSelector.tsx

```diff
@@ -0,0 +1,84 @@
+import { Dispatch, SetStateAction, useLayoutEffect, useState } from "react";
+
+import Alert from "@mui/material/Alert";
+import CircularProgress from "@mui/material/CircularProgress";
+import FormControl from "@mui/material/FormControl";
+import InputLabel from "@mui/material/InputLabel";
+import MenuItem from "@mui/material/MenuItem";
+import Select, { SelectChangeEvent } from "@mui/material/Select";
+
+type props = {
+  value: string;
+  setValue: Dispatch<SetStateAction<string>>;
+};
+
+const DatatypeSelector = ({ value, setValue }: props) => {
+  const [error, setError] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [items, setItems] = useState([]);
+
+  const handleChange = (event: SelectChangeEvent) => {
+    setValue(event.target.value as string);
+  };
+
+  useLayoutEffect(() => {
+    setIsLoading(true);
+    fetch(`${process.env.REACT_APP_BACKEND}/data_types`)
+      .then((res) => {
+        if (res.ok) {
+          return res.json();
+        }
+        return Promise.reject(res.statusText);
+      })
+      .then((result) => {
+        setIsLoading(false);
+        setItems(result);
+      })
+      .catch((error) => {
+        setIsLoading(false);
+        setError(error);
+      });
+  }, []);
+
+  if (isLoading) {
+    return (
+      <div className="m-4">
+        <CircularProgress />
+      </div>
+    );
+  }
+
+  if (error) {
+    return (
+      <div className="m-4">
+        <Alert severity="error">
+          Encountered some technical difficulties, please try again later!
```

Comment:
Would it be practical to display a more specific error message to the user?
e.g in line 39:
```
setError(error.message);
```
then print out the error here
```
<Alert severity="error">
    Encountered some technical difficulties, please try again later!
    Error: {error}
</Alert>
```

Does this sound sensible @andrewyatz ? or is it better to keep it as it is for now?

Reply:
1. useLayoutEffect triggers as soon as the component is called even before the browser starts the DOM painting so this will help to ensure that we have more time to fetch the required data and thereby reducing the chances of user seeing any updates happening on the DOM after the component has rendered.

2. Yes I did think of that scenario and tested that too, actually, it will be
SELECT * FROM gene WHERE species='homo_sapiens' 
  AND gene_stable_id='' AND gene_chr='' 
  AND ((gene_start >= AND gene_start <= ) 
  OR (gene_end >= AND gene_end <= ) 
  OR (gene_start <= AND gene_end >= ));

I tried handling this by making the textboxes required but that will also require me to use Form and submit the request via that, because of lack of time I skipped that for now but yeah this is something that's good to have from UX point of view but considering the main focus was to make the filter work dynamically so I prioritised accordingly, will try to accommodate these corner cases in future as time permits

3. AWS does not provide any way to perform such validations before generating the query ID, it can only tell the issues after the query has started processing and a FAILED state indicates there is a issue with the query submitted


In this particular scenario chances of backend giving any error is nil, only possible scenario where user will see an error is when there is a network error, on top of that these values are coming from cache so I have kept just this error statement because same error can come when there is a network issue either at server or client side

Reply:
1. My only concern regarding `useLayoutEffect` is that being a synchronous hook causes the display to not render anything until the hook is done running. If fetching data is slow the components won’t render anything until the fetching operation is done.
2. That makes sense, I do agree, these special cases can be tackled later if time permits
3. I see, thanks, it's good to keep that in mind

The reason why I mentioned the error is that I got an error when deploying the UI locally because of CROS
```
Access to fetch at 'http://IP/data_types' from origin 'http://localhost:3000' 
has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present 
on the requested resource. If an opaque response serves your needs, set the request's 
mode to 'no-cors' to fetch the resource with CORS disabled.
```